pkg/test/integration/server.go

```diff
@@ -0,0 +1,202 @@
+// server.
+func NewCleaner() OperationFn {
+	return func(e *envtest.Environment, c client.Client) error {
+		cs, err := clientset.NewForConfig(e.Config)
+		if err != nil {
+			return errors.Wrap(err, errCleanup)
+		}
+		var crds []*apiextensionsv1beta1.CustomResourceDefinition
+		for _, path := range e.CRDDirectoryPaths {
+			crd, err := readCRDs(path)
+			if err != nil {
+				return errors.Wrap(err, errCleanup)
+			}
+			crds = append(crds, crd...)
+		}
+
+		for _, crd := range crds {
+			if err := cs.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(crd.Name, nil); err != nil {
+				return errors.Wrap(err, errCleanup)
+			}
+		}
+		return nil
+	}
+}
+
+// An Option configures an Athodyd Config.
+type Option func(*Config)
+
+// WithBuilder sets a custom builder function for an Athodyd Config.
+func WithBuilder(builder OperationFn) Option {
+	return func(c *Config) {
+		c.Builder = builder
+	}
+}
+
+// WithCleaner sets a custom cleaner function for an Athodyd Config.
+func WithCleaner(cleaner OperationFn) Option {
+	return func(c *Config) {
+		c.Cleaner = cleaner
+	}
+}
+
+// WithCRDDirectoryPaths sets custom CRD locations for an Athodyd Config.
+func WithCRDDirectoryPaths(crds ...string) Option {
+	return func(c *Config) {
+		c.CRDDirectoryPaths = crds
+	}
+}
+
+// WithManagerOptions sets custom options for the manager configured by Athodyd
+// Config.
+func WithManagerOptions(m manager.Options) Option {
+	return func(c *Config) {
+		c.ManagerOptions = m
+	}
+}
+
+func defaultConfig() *Config {
+	t, err := time.ParseDuration(syncPeriod)
+	if err != nil {
+		panic(err)
+	}
+
+	return &Config{
+		CRDDirectoryPaths: []string{},
+		Builder:           NewBuilder(),
+		Cleaner:           NewCleaner(),
+		ManagerOptions:    manager.Options{SyncPeriod: &t},
+	}
+}
+
+// Manager wraps a controller-runtime manager with additional functionality.
+type Manager struct {
+	manager.Manager
+	channel chan struct{}
+	env     *envtest.Environment
+	client  client.Client
+	c       *Config
+}
+
+// New creates a new Athodyd Manager.
+func New(cfg *rest.Config, o ...Option) (*Manager, error) {
+	var useExisting bool
+	if cfg != nil {
+		useExisting = true
+	}
+
+	c := defaultConfig()
+	for _, op := range o {
+		op(c)
+	}
+
+	e := &envtest.Environment{
+		CRDDirectoryPaths:  c.CRDDirectoryPaths,
+		Config:             cfg,
+		UseExistingCluster: &useExisting,
+	}
+
+	cfg, err := e.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := client.New(cfg, client.Options{})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := c.Builder(e, client); err != nil {
+		return nil, err
+	}
+
+	mgr, err := manager.New(cfg, c.ManagerOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	channel := make(chan struct{})
+	return &Manager{mgr, channel, e, client, c}, nil
+}
+
+// Run starts a controller-runtime manager with Athodyd's channel.
+func (m *Manager) Run() {
+	go func() {
+		if err := m.Start(m.channel); err != nil {
```

Comment:
Do you think it would make sense for `channel` to have a more explicit name? `channel` does not do much to communicate its purpose. Should we embed a stop channel at all, rather than relying on the caller to pass one in? I imagine there is a reason `envtest` was designed such that the caller passes in a channel.

Reply:
I think calling it `stop` would be appropriate. `envtest` doesn't actually do any manager setup, so there is not channel being passed in. My initial thinking was that since this can only ever be an empty struct channel (a.k.a. a "signal channel") I don't see any enhancement from an end user perspective in letting someone pass in their own channel. However, if a user wanted to test stopping the manager (and thus any registered controllers), then continuing with any tests, it may be useful to have more fine-grained control of the channel. I don't see this as a major use-case (if resources want to be tested without controllers running then a separate test can be created that does not register the controllers), but I could be wrong when we start using this more. If it is all the same to you, I would prefer to implement in a later PR as needed :)